README.md
<h1>LG Smart TV remote switcher for 3D content</h1>
This addon turns your LG Smart TV into 3D mode automatically over the network if tagged '3D' content is playing (and your graphics adapter can't do this). For this, the LG TV device must be integrated in your home network properly. This addon supports LG TVs of the old 2011 series and the newer 2012 models too.

changelog.txt
Changelog

0.1.20
- smaller changes, bump version due repo change

0.1.15
- Bugfix

0.1.14
- Log messages optimized

0.1.13
- Changing language files to transifex .po
- Minor changes

0.1.12
- Minor changes

0.1.11
- Reload settings if setup was changed while processing
- Improved handling if a device is disconnected

0.1.10
- Exception handling on SystemExit implemented

0.1.9
- Command sequence of 3D Mode now depends from selected/current view mode instead of filename convention
  This prevents from switching to view modes defined by filename if user select other view mode as suggested

0.1.8
- Addon splitted into script module and service module for better compatibility

0.1.7
- lower delay values reimplemented

0.1.4
- smaller improvements

0.1.3
- Key delay redefined
- improved communication handling

0.1.2
- Default code sequences redefined

0.1.0
  - Second release
  - extended settings
  - user defined key codes
  - adjustable key delay

0.0.2
  - Improved setup handling

0.0.1
  - Initial version


scan.py
import xbmc, xbmcaddon
import os
import sys
from resources.lib import interface
from resources.lib import tools

__addon__ = xbmcaddon.Addon()
__addonname__ = __addon__.getAddonInfo('name')
__addonID__ = __addon__.getAddonInfo('id')
__path__ = __addon__.getAddonInfo('path')
__version__ = __addon__.getAddonInfo('version')
__LS__ = __addon__.getLocalizedString

__IconConnected__ = xbmc.translatePath(os.path.join( __path__,'resources', 'media', 'ok.png'))
__IconError__ = xbmc.translatePath(os.path.join( __path__,'resources', 'media', 'fail.png'))
__IconDefault__ = xbmc.translatePath(os.path.join( __path__,'resources', 'media', 'default.png'))

try:
    if sys.argv[1] == 'scan':
        tools.notifyLog("Scanning for LG Smart TV Devices...", level=xbmc.LOGDEBUG)

        _host = None if __addon__.getSetting('lg_host') == '' else __addon__.getSetting('lg_host')
        _port = 8080 if __addon__.getSetting('lg_port') == '' else int(__addon__.getSetting('lg_port'))
        Remote = interface.Interface(host=_host, port=_port, protocol=None)
        _host = Remote.host
        _protocol = Remote._protocol
        tools.notifyLog('Device (IP %s protocol %s) found' % (_host, _protocol.upper()), level=xbmc.LOGDEBUG)
        #
        # input pairing key if not exists
        #
        _pairing_key = None if __addon__.getSetting('lg_pairing_key') == '' else __addon__.getSetting('lg_pairing_key')
        if _pairing_key is None:
            kb = xbmc.Keyboard('', __LS__(30030))
            kb.doModal()
            if kb.isConfirmed() and kb.getText() != '':  _pairing_key = kb.getText()

        _conn = Remote.get_session_id(_pairing_key)
        if _conn:
            tools.notifyLog('Session with ID %s established' % (Remote.session_id), level=xbmc.LOGDEBUG)
            # we are ready

            __addon__.setSetting('lg_host', _host)
            __addon__.setSetting('lg_protocol', _protocol.upper())
            __addon__.setSetting('lg_pairing_key', _pairing_key)
            if tools.dialogYesNo(__LS__(30031) % (_host, _protocol.upper())):
                tools.notifyLog('Restart Application', level=xbmc.LOGDEBUG)
                xbmc.executebuiltin('RestartApp')
            else:
                tools.notifyLog('User decided not to restart application...', level=xbmc.LOGDEBUG)
        else:
            tools.notifyLog('Session not established. Try again.', xbmc.LOGERROR)
            tools.dialogOSD(__LS__(30032))

except interface.Interface.LGinNetworkNotFoundException:
    tools.notifyLog('LG Devices not found in network.', level=xbmc.LOGERROR)
    tools.dialogOSD( __LS__(30050))
except interface.Interface.LGProtocolWebOSException:
    tools.notifyLog('Device use WebOS on port 3000. Not supported.', level=xbmc.LOGERROR)
    tools.dialogOSD(__LS__(30051))
except interface.Interface.LGProtocollNotAcceptedException:
    tools.notifyLog('Protocol not supported.', level=xbmc.LOGERROR)
    tools.dialogOSD(__LS__(30052))
except interface.Interface.NoConnectionToHostException:
    tools.notifyLog('No connection to host.', level=xbmc.LOGERROR)
    tools.dialogOSD(__LS__(30053) % (_host))
except IndexError:
    tools.notifyLog('Calling this script outside of setup not allowed', xbmc.LOGERROR)
    tools.dialogOSD(__LS__(30055))

resources/lib/tools.py
import xbmc
import xbmcgui
import xbmcaddon
import os

__addon__ = xbmcaddon.Addon()
__addonname__ = __addon__.getAddonInfo('name')
__addonID__ = __addon__.getAddonInfo('id')
__path__ = __addon__.getAddonInfo('path')
__version__ = __addon__.getAddonInfo('version')
__LS__ = __addon__.getLocalizedString

__IconConnected__ = xbmc.translatePath(os.path.join( __path__,'resources', 'media', 'ok.png'))
__IconError__ = xbmc.translatePath(os.path.join( __path__,'resources', 'media', 'fail.png'))
__IconDefault__ = xbmc.translatePath(os.path.join( __path__,'resources', 'media', 'default.png'))

OSD = xbmcgui.Dialog()

def notifyOSD(header, message, icon=__IconDefault__):
    OSD.notification(header.encode('utf-8'), message.encode('utf-8'), icon)

def dialogOSD(message, header=__addonname__):
    OSD.ok(header.encode('utf-8'), message.encode('utf-8'))

def dialogYesNo(message, header=__addonname__):
    return OSD.yesno(header.encode('utf-8'), message.encode('utf-8'))

def notifyLog(message, level=xbmc.LOGNOTICE):
    xbmc.log('[%s %s] %s' % (__addonID__, __version__, message.encode('utf-8')), level)

resources/lib/interface.py
#!/usr/bin/env python
# encoding: utf-8

import re
import logging
import socket
import time
import tools

import xml.etree.ElementTree as etree
import httplib
import xbmc

# KEY_PAIRING='SMYBYA' # LG at Home
# KEY_PAIRING='045855' # LG at Work

class KeyInputError(Exception):
    pass

class Interface(object):

    _xml_version_string = '<?xml version="1.0" encoding="utf-8"?>'
    _headers = {'Content-Type': 'application/atom+xml'}
    _highest_key_input_for_protocol = {'hdcp': 255, 'roap': 1024}

    class LGinNetworkNotFoundException(Exception): pass
    class LGProtocolWebOSException(Exception): pass
    class LGProtocollNotAcceptedException(Exception): pass
    class NoConnectionToHostException(Exception): pass

    def __init__(self, host=None, port=8080, protocol=None):

        self.port = int(port)
        self.host = host
        if host == None: self.host = self.getip()

        self._protocol = protocol
        if protocol == None:
            self._protocol = self.auto_detect_accepted_protocol()

        self._pairing_key = None
        self.session_id = None

    def getip(self):
        if self.host: return self.host
        strngtoXmit = 'M-SEARCH * HTTP/1.1\r\nHOST: 239.255.255.250:1900\r\n' + \
            'MAN: "ssdp:discover"\r\nMX: 2\r\nST: urn:schemas-upnp-org:device:MediaRenderer:1\r\n\r\n'

        bytestoXmit = strngtoXmit.encode()
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        sock.settimeout(3)
        found = False
        i = 0

        while not found and i < 5:
            try:
                sock.sendto(bytestoXmit, ('239.255.255.250', 1900))
                gotbytes, addressport = sock.recvfrom(512)
                gotstr = gotbytes.decode()
                if re.search('LG', gotstr):
                    tools.notifyLog('Returned: %s' % gotstr, level=xbmc.LOGDEBUG)
                    self.host, self.port = addressport
                    tools.notifyLog('Found device: %s' % self.host, level=xbmc.LOGDEBUG)
                    found = True
                    break
                i += 1
                time.sleep(1)
            except:
                pass
        sock.close()

        if not found: raise self.LGinNetworkNotFoundException('LG TV not found')
        tools.notifyLog("Using device: %s over transport protocol: %s" % (self.host, self.port), level=xbmc.LOGDEBUG)
        return self.host

    def auto_detect_accepted_protocol(self):
        if self._doesServiceExist(3000):
            tools.notifyLog("Device use WebOS on Port 3000. Not supported.")
            raise self.LGProtocolWebOSException("WebOS not supported.")

        req_key_xml_string = self._xml_version_string + '<auth><type>AuthKeyReq</type></auth>'
        tools.notifyLog("Try to detect accepted protocols", level=xbmc.LOGDEBUG)

        try:
            for protocol in self._highest_key_input_for_protocol:
                tools.notifyLog("Testing protocol: %s" % (protocol), level=xbmc.LOGDEBUG)
                conn = httplib.HTTPConnection(self.host, port=self.port, timeout=3)
                conn.request("POST", "/%s/api/auth" % (protocol), req_key_xml_string, headers=self._headers)
                http_response = conn.getresponse()
                tools.notifyLog("Got response: %s" % (http_response.reason), level=xbmc.LOGDEBUG)
                if http_response.reason == 'OK':
                    self._protocol = protocol
                    tools.notifyLog("Using protocol: %s" % (self._protocol), level=xbmc.LOGDEBUG)
                    return self._protocol
            raise self.LGProtocollNotAcceptedException("No accepted protocol found.")
        except:
            raise self.NoConnectionToHostException("No connection to host %s" % (self.host))

    def display_key_on_screen(self):
        conn = httplib.HTTPConnection(self.host, port=self.port)
        req_key_xml_string = self._xml_version_string + '<auth><type>AuthKeyReq</type></auth>'
        tools.notifyLog("Request device to show key on screen.", level=xbmc.LOGDEBUG)
        conn.request('POST', '/%s/api/auth' % (self._protocol), req_key_xml_string, headers=self._headers)
        http_response = conn.getresponse()
        tools.notifyLog("Device response was: %s" % (http_response.reason), level=xbmc.LOGDEBUG)
        if http_response.reason != "OK": raise Exception("Network error: %s" % (http_response.reason))

        return http_response.reason

    def get_session_id(self, pairing_key):
        if not pairing_key: return False

        self._pairing_key = pairing_key
        tools.notifyLog("Trying paring key: %s" % (self._pairing_key), level=xbmc.LOGDEBUG)
        pair_cmd_xml_string = self._xml_version_string + '<auth><type>AuthReq</type><value>' + \
            self._pairing_key + '</value></auth>'
        try:
            conn = httplib.HTTPConnection(self.host, port=self.port, timeout=3)
            conn.request('POST', '/%s/api/auth' % (self._protocol), pair_cmd_xml_string, headers=self._headers)
            http_response = conn.getresponse()
            if http_response.reason != 'OK': return False

            tree = etree.XML(http_response.read())
            self.session_id = tree.find('session').text
            tools.notifyLog("Session ID is %s" % (self.session_id), level=xbmc.LOGDEBUG)
            if len(self.session_id) < 8: return False

            return self.session_id
        except socket.timeout:
            raise self.NoConnectionToHostException("No connection to host %s" % (self.host))
        except socket.error:
            raise self.NoConnectionToHostException("No connection to host %s" % (self.host))

    def handle_key_input(self, cmdcode):
        highest_key_input = self._highest_key_input_for_protocol[self._protocol]
        try:
            if 0 > int(cmdcode) or int(cmdcode) > highest_key_input:
                raise KeyInputError("Key input %s is not supported." % (cmdcode))
        except ValueError:
            raise KeyInputError("Key input %s is not a number" % (cmdcode))
        if not self.session_id: raise Exception("No valid session key available.")

        command_url_for_protocol = {
            'hdcp': '/%s/api/dtv_wifirc' % (self._protocol),
            'roap': '/%s/api/command' % (self._protocol),
        }

        key_input_xml_string = self._xml_version_string + '<command><session>' + self.session_id \
            + '</session><type>HandleKeyInput</type><value>' + cmdcode + '</value></command>'
        conn = httplib.HTTPConnection(self.host, port=self.port)
        conn.request('POST', command_url_for_protocol[self._protocol], key_input_xml_string, headers=self._headers)
        return conn.getresponse().reason

    def _doesServiceExist(self, port):
        try:
            tools.notifyLog("Checking port %s" % (port), level=xbmc.LOGDEBUG)
            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            s.settimeout(1)
            s.connect((self.host, port))
            s.close()
        except:
            return False
        return True


resources/lib/lgcommander.py
#!/usr/bin/env python3
# encoding: utf-8

# Todo:
# Use the same TCP connection.

__version__ = '0.5'

import re
import logging
import socket

import xml.etree.ElementTree as etree
import httplib

KEY_IDX_3D=400
KEY_IDX_ARROW_DOWN=2
KEY_IDX_ARROW_LEFT=3
KEY_IDX_ARROW_RIGHT=4
KEY_IDX_ARROW_UP=1
KEY_IDX_BACK=23
KEY_IDX_BLUE=29
KEY_IDX_BTN_1=5
KEY_IDX_BTN_2=6
KEY_IDX_BTN_3=7
KEY_IDX_BTN_4=8
KEY_IDX_CH_DOWN=28
KEY_IDX_CH_UP=27
KEY_IDX_ENTER=20
KEY_IDX_EXIT=412
KEY_IDX_EXTERNAL_INPUT=47
KEY_IDX_GREEN=30
KEY_IDX_HOME=21
KEY_IDX_MUTE=26
KEY_IDX_MYAPPS=417
KEY_IDX_NETCAST=408
KEY_IDX_PAUSE=34
KEY_IDX_PLAY=33
KEY_IDX_POWER_OFF=1
KEY_IDX_PREV_CHANNEL=403
KEY_IDX_RED=31
KEY_IDX_STOP=35
KEY_IDX_VOL_DOWN=25
KEY_IDX_VOL_UP=24
KEY_IDX_YELLOW=32

'''
from tkinter import Tk, Toplevel, Entry, Button, Label

class MyDialog:

    def __init__(self, parent, dialog_msg):
        self.user_string = None

        top = self.top = Toplevel(parent)
        Label(top, text=dialog_msg, justify='left').pack()
        self.e = Entry(top)
        self.e.pack(padx=5)
        self.e.focus_set()
        b = Button(top, text='Ok', command=self.ok)
        b.pack(pady=5)
        top.bind('<Return>', self.ok)
        top.title("Lg Commander")
        top.geometry("410x280+10+10")

    def ok(self, dummy=None):
        self.user_string = self.e.get()
        self.top.destroy()
'''

class KeyInputError(Exception):
    pass

class LgRemote:

    _xml_version_string = '<?xml version="1.0" encoding="utf-8"?>'
    _headers = {'Content-Type': 'application/atom+xml'}
    _highest_key_input_for_protocol = {'hdcp': 255, 'roap': 1024}

    def __init__(self, host=None, port=8080, protocol=None):

        self.port = int(port)
        self.host = host
        if host == None: self.getip()

        if protocol == None:
            self.auto_detect_accepted_protocol()
        else:
            self._protocol = protocol

        self._pairing_key = None
        self._session_id = None

    def getip(self):
        if self.host: return self.host
        strngtoXmit = 'M-SEARCH * HTTP/1.1' + '\r\n' + 'HOST: 239.255.255.250:1900' + '\r\n' + \
            'MAN: "ssdp:discover"' + '\r\n' + 'MX: 2' + '\r\n' + \
            'ST: urn:schemas-upnp-org:device:MediaRenderer:1' + '\r\n' + '\r\n'

        bytestoXmit = strngtoXmit.encode()
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        sock.settimeout(3)
        found = False
        gotstr = 'notyet'
        i = 0

        while not found and i <= 5:
            try:
                sock.sendto(bytestoXmit, ('239.255.255.250', 1900))
                gotbytes, addressport = sock.recvfrom(512)
                gotstr = gotbytes.decode()
                if re.search('LG', gotstr):
                    logging.debug('Returned: %s' % gotstr)
                    self.host, port = addressport
                    logging.debug('Found device: %s' % self.host)
                    found = True
                    break
                i += 1
            except:
                pass
        sock.close()

        if not found: raise socket.error("LG TV not found.")
        logging.info("Using device: %s over transport protocol: %s" % (self.host, self.port))
        return self.host

    def auto_detect_accepted_protocol(self):
        req_key_xml_string = self._xml_version_string + '<auth><type>AuthKeyReq</type></auth>'
        logging.debug("Detecting accepted protocol.")
        if self._doesServiceExist(3000):
            raise Exception("Protocol not supported. See https://github.com/ypid/lgcommander/issues/1")
        try:
            for protocol in self._highest_key_input_for_protocol:
                logging.debug("Testing protocol: %s" % (protocol))
                conn = httplib.HTTPConnection(self.host, port=self.port, timeout=5)
                conn.request("POST", "/%s/api/auth" % (protocol), req_key_xml_string, headers=self._headers)
                http_response = conn.getresponse()
                logging.debug("Got response: %s" % (http_response.reason))
                if http_response.reason == 'OK':
                    self._protocol = protocol
                    logging.debug("Using protocol: %s" % (self._protocol))
                    return self._protocol
            raise Exception("No accepted protocol found.")
        except:
            raise socket.error("No connection to host %s" % (self.host))

    def display_key_on_screen(self):
        conn = httplib.HTTPConnection(self.host, port=self.port)
        req_key_xml_string = self._xml_version_string + '<auth><type>AuthKeyReq</type></auth>'
        logging.debug("Request device to show key on screen.")
        conn.request('POST', '/%s/api/auth' % (self._protocol), req_key_xml_string, headers=self._headers)
        http_response = conn.getresponse()
        logging.debug("Device response was: %s" % (http_response.reason))
        if http_response.reason != "OK": raise Exception("Network error: %s" % (http_response.reason))

        return http_response.reason

    def get_session_id(self, paring_key):
        if not paring_key: return None

        self._pairing_key = paring_key
        logging.debug("Trying paring key: %s" % (self._pairing_key))
        pair_cmd_xml_string = self._xml_version_string + '<auth><type>AuthReq</type><value>' + \
            self._pairing_key + '</value></auth>'
        conn = httplib.HTTPConnection(self.host, port=self.port)
        conn.request('POST', '/%s/api/auth' % (self._protocol), pair_cmd_xml_string, headers=self._headers)
        http_response = conn.getresponse()
        if http_response.reason != 'OK': return None

        tree = etree.XML(http_response.read())
        self._session_id = tree.find('session').text
        logging.debug("Session ID is %s" % (self._session_id))
        if len(self._session_id) < 8: raise Exception("Could not get Session Id: %s" % (self._session_id))

        return self._session_id

    def handle_key_input(self, cmdcode):
        highest_key_input = self._highest_key_input_for_protocol[self._protocol]
        try:
            if 0 > int(cmdcode) or int(cmdcode) > highest_key_input:
                raise KeyInputError("Key input %s is not supported." % (cmdcode))
        except ValueError:
            raise KeyInputError("Key input %s is not a number" % (cmdcode))
        if not self._session_id: raise Exception("No valid session key available.")

        command_url_for_protocol = {
            'hdcp': '/%s/api/dtv_wifirc' % (self._protocol),
            'roap': '/%s/api/command' % (self._protocol),
        }

        logging.debug("Executing command: %s" % (cmdcode))
        key_input_xml_string = self._xml_version_string + '<command><session>' + self._session_id \
            + '</session><type>HandleKeyInput</type><value>' + cmdcode + '</value></command>'
        conn = httplib.HTTPConnection(self.host, port=self.port)
        conn.request('POST', command_url_for_protocol[self._protocol], key_input_xml_string, headers=self._headers)
        return conn.getresponse()

    def _doesServiceExist(self, port):
        try:
            logging.debug("Trying to connect to port %s/tcp" % (port))
            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            s.settimeout(1)
            s.connect((self.host, port))
            s.close()
        except:
            return False
        return True

def main():  # {{{
    """Execute module in command line mode."""

    def get_pairing_key_from_user(lg_remote):
        lg_remote.display_key_on_screen()
        '''
        root = Tk()
        root.withdraw()
        dialog_msg = "Please enter the pairing key\nyou see on your TV screen\n"
        dialog = MyDialog(root, dialog_msg)
        root.wait_window(dialog.top)
        session_id = dialog.user_string
        dialog.top.destroy()
        '''
        session_id='045855'
        return session_id

    args = ArgumentParser(
        description="Control your Smart Lg TV with your PC",
    )
    args.add_argument(
        '-V',
        '--version',
        action='version',
        version='%(prog)s {version}'.format(version=__version__)
    )
    args.add_argument(
        '-H',
        '--host',
        default='scan',
        help="IP address or FQDN of device."
        + " Use the special value \"scan\" for a multicast request for TVs in your LAN."
        + " \"scan\" will also be used if this parameter was omitted."
    )
    args.add_argument(
        '-p',
        '--port',
        default='8080',
        help="TCP port (default is 8080)."
    )
    args.add_argument(
        '-P',
        '--protocol',
        choices=['roap', 'hdcp'],
        default=None,
        help="Protocol to use."
        + " Currently ROAP and HDCP are supported."
        + " Default is to auto detect the correct one.",
    )
    args.add_argument(
        '-k',
        '--pairing-key',
        help="Pairing key of your TV."
        + " This key is shown on request on the screen"
        + " and does only change if you factory reset your TV."
    )
    args.add_argument(
        '-c',
        '--command',
        help="Send just a single command and exit."
    )
    user_parms = args.parse_args()

    logging.basicConfig(
        format='# %(levelname)s: %(message)s',
        level=logging.DEBUG,
        # level=logging.INFO,
    )

    try:
        lg_remote = LgRemote(
            host=None if user_parms.host == 'scan' else user_parms.host,
            port=user_parms.port,
            protocol=user_parms.protocol)
    except socket.error as error:
        raise SystemExit(error)

    if user_parms.pairing_key:
        logging.debug("Pairing key from user {}".format(user_parms.pairing_key))
        lg_remote.get_session_id(user_parms.pairing_key)
    while not lg_remote._session_id:
        logging.debug("No valid pairing key available. Showing key on TV screen...")
        lg_remote.get_session_id(get_pairing_key_from_user(lg_remote))

    dialog_msg = "\nSession ID: " + str(lg_remote._session_id) + "\n"
    dialog_msg += "Paring key: " + str(lg_remote._pairing_key) + "\n"
    dialog_msg += "Success in establishing command session\n"
    dialog_msg += "_" * 64 + "\n\n"
    dialog_msg += "Enter command code i.e. a number between 0 and 255\n"
    dialog_msg += "Enter a number greater than 255 to quit.\n"
    dialog_msg += "Some useful codes:\n"
    dialog_msg += "for EZ_ADJUST menu    enter   255 \n"
    dialog_msg += "for IN START menu     enter   251 \n"
    dialog_msg += "for Installation menu enter   207 \n"
    dialog_msg += "for POWER_ONLY mode   enter   254 \n"
    dialog_msg += "Warning: do not enter 254 if you \n"
    dialog_msg += "do not know what POWER_ONLY mode is. "

    logging.info(dialog_msg)

    if user_parms.command:
        lg_remote.handle_key_input(user_parms.command)
        raise SystemExit()

    '''
    while True:
        root = Tk()
        root.withdraw()
        dialog = MyDialog(root, dialog_msg)
        root.wait_window(dialog.top)
        try:
            lg_remote.handle_key_input(dialog.user_string)
        except KeyInputError:
            logging.debug("Terminate on user requested.")
            break
    '''

if __name__ == '__main__':
    from argparse import ArgumentParser

    main()
# }}}


default.py
import xbmc, xbmcaddon
import os
import json
import re
from resources.lib import interface
from resources.lib import tools

# regular expressions for 3d detection (filenames)
# see http://kodi.wiki/view/Advancedsettings.xml#video or
# http://kodi.wiki/view/3D

__pattern3D__ =   '[-. _]3d[-. _]'
__patternSBS__ =  '[-. _]h?sbs[-. _]'
__patternTAB__ =  '[-. _]h?tab[-. _]'

# Key sequences ROAP/HDCP
__mode3DSBS_on__ =     {'roap': ['400', '15', '20'], 'hdcp': ['220', '68', '6', '68']}
__mode3DSBS_off__ =    {'roap': ['400', '20', '400', '14', '20', '400'], 'hdcp': ['220', '68', '220', '68', '7', '68', '220', '68']}
__mode3DTAB_on__ =     {'roap': ['400', '15', '15', '20'], 'hdcp': ['220', '68', '6', '6', '68']}
__mode3DTAB_off__ =    {'roap': ['400', '20', '400', '14', '14', '20', '400'], 'hdcp': ['220', '68', '220', '68', '7', '7', '68', '220', '68']}

__mode3D_on__ = None
__mode3D_off__ = None

WAIT_FOR_MODE_SELECT = 60
WAIT_FOR_NEXT_CONNECT = 30
POLL_INTERVAL = 1

__addon__ = xbmcaddon.Addon()
__addonname__ = __addon__.getAddonInfo('name')
__addonID__ = __addon__.getAddonInfo('id')
__path__ = __addon__.getAddonInfo('path')
__version__ = __addon__.getAddonInfo('version')
__LS__ = __addon__.getLocalizedString

__IconConnected__ = xbmc.translatePath(os.path.join( __path__,'resources', 'media', 'ok.png'))
__IconError__ = xbmc.translatePath(os.path.join( __path__,'resources', 'media', 'fail.png'))
__IconDefault__ = xbmc.translatePath(os.path.join( __path__,'resources', 'media', 'default.png'))

class Monitor(xbmc.Monitor):

    def __init__(self):

        xbmc.Monitor.__init__(self)
        self.settingsChanged = False
        self.abortRequested = False

    def onSettingsChanged(self):
        self.settingsChanged = True

    def onAbortRequested(self):
        self.abortRequested = True

class Service(xbmc.Player):

    def __init__(self):

        xbmc.Player.__init__(self)
        self.Mon = Monitor()
        self.Remote = None
        self.sessionEstablished = False
        self.getSettings()

        self.isPlaying3D = None
        self.mode3D = 'OFF'

    def getSettings(self, init=True):
        if init: self.readSettings()
        if self.lg_host is not None:
            if self.Remote is None:
                try:
                    self.Remote = interface.Interface(self.lg_host, self.lg_port, self.lg_protocol)
                    if not self.Remote.session_id:
                        self.Remote.get_session_id(self.lg_pairing_key)
                        self.sessionEstablished = True
                        tools.notifyLog('Session established. Using session id %s.' % (self.Remote.session_id), level=xbmc.LOGDEBUG)
                except self.Remote.NoConnectionToHostException:
                    self.sessionEstablished = False
                    self.Remote = None
                    tools.notifyLog('No connection to host on %s' % (self.lg_host), level=xbmc.LOGERROR)
                    if init: tools.notifyOSD(__addonname__, __LS__(30054), icon=__IconError__)

    def readSettings(self):
        self.lg_host = __addon__.getSetting('lg_host')
        self.lg_host = None if self.lg_host == '' else self.lg_host
        self.lg_port = __addon__.getSetting('lg_port')
        self.lg_protocol = __addon__.getSetting('lg_protocol')
        self.lg_protocol = None if self.lg_protocol == __LS__(30017) else self.lg_protocol.lower()
        self.lg_pairing_key = __addon__.getSetting('lg_pairing_key')

        self.lg_key_delay = int(re.match('\d+', __addon__.getSetting('lg_delay')).group())
        self.lg_own_seqs_enabled = True if __addon__.getSetting('use_own_seq').upper() == 'TRUE' else False
        self.lg_seq_3D_on = ' '.join(__addon__.getSetting('lg_3D_on').replace(',',' ').split()).split()
        self.lg_seq_3D_off = ' '.join(__addon__.getSetting('lg_3D_off').replace(',',' ').split()).split()

        self.Mon.settingsChanged = False
        tools.notifyLog('Settings reloaded', level=xbmc.LOGDEBUG)

    def getStereoscopicMode(self):
        mode = {"off": 'OFF', "split_vertical": 'SBS', "split_horizontal": 'TAB',
                 "row_interleaved": 'INTERLEAVE', "hardware_based": 'HW', "anaglyph_cyan_red": 'CR',
                 "anaglyph_green_magenta": 'GM', "monoscopic": 'MONO'}
        query = {
                "jsonrpc": "2.0",
                "method": "GUI.GetProperties",
                "params": {"properties": ["stereoscopicmode"]},
                "id": 1
                }
        _poll = WAIT_FOR_MODE_SELECT
        while _poll > 0:
            try:
                res = json.loads(xbmc.executeJSONRPC(json.dumps(query, encoding='utf-8')))
                if 'result' in res and 'stereoscopicmode' in res['result']:
                    res = res['result']['stereoscopicmode'].get('mode')
                    if self.mode3D != mode[res]:
                        tools.notifyLog('Stereoscopic mode has changed to %s' % (mode[res]))
                        self.mode3D = mode[res]
                        return True
                    _poll -= 1
                    xbmc.sleep(1000)
                else:
                    break
            except SystemExit:
                tools.notifyLog('System will terminate this script, closing it.', level=xbmc.LOGERROR)
                break
        tools.notifyLog('Could not determine steroscopic mode', level=xbmc.LOGERROR)
        return False

    def sendCommand(self, sequence, own_sequence):
        try:
            self.getSettings()
            if self.lg_own_seqs_enabled:
                sequence = own_sequence
                tools.notifyLog('Sending user sequence %s' % (sequence), level=xbmc.LOGDEBUG)

            for code in sequence:
                if self.Remote.session_id is None: self.Remote.get_session_id(self.lg_pairing_key)
                # let smart models time for response ;)
                xbmc.sleep(self.lg_key_delay)
                tools.notifyLog('%s msec delayed, sending keycode %s. Response: %s.' % (self.lg_key_delay, code, self.Remote.handle_key_input(code)), level=xbmc.LOGDEBUG)

        except self.Remote.NoConnectionToHostException:
            self.sessionEstablished = False
            tools.notifyLog('No connection to host on %s' % (self.lg_host), level=xbmc.LOGERROR)

    def onPlayBackStarted(self):
        if self.isPlayingVideo() and self.lg_protocol is not None:
            _file = os.path.basename(self.getPlayingFile().decode('utf-8'))
            if re.search(__pattern3D__, _file, re.IGNORECASE):
                if self.getStereoscopicMode():
                    if self.mode3D == 'SBS': __mode3D_on__ = __mode3DSBS_on__
                    elif self.mode3D == 'TAB': __mode3D_on__ = __mode3DTAB_on__
                    else:
                        return

                tools.notifyLog('Playing \'%s\'' % (_file), level=xbmc.LOGDEBUG)
                tools.notifyLog('sending sequence for 3D %s' % (self.mode3D), level=xbmc.LOGDEBUG)
                self.sendCommand(__mode3D_on__[self.lg_protocol], self.lg_seq_3D_on)
                self.isPlaying3D = True

    def onPlayBackStopped(self):
        _currentMode = self.mode3D
        if self.getStereoscopicMode() and self.isPlaying3D and self.mode3D == 'OFF':
            tools.notifyLog('Turn 3D %s mode off' % (_currentMode), level=xbmc.LOGDEBUG)
            if _currentMode == 'SBS':
                __mode3D_off__ = __mode3DSBS_off__
            elif _currentMode == 'TAB':
                __mode3D_off__ = __mode3DTAB_off__
            else:
                return

            self.sendCommand(__mode3D_off__[self.lg_protocol], self.lg_seq_3D_off)
            self.isPlaying3D = False

    def onPlayBackEnded(self):
        _currentMode = self.mode3D
        if self.getStereoscopicMode() and self.isPlaying3D and self.mode3D == 'OFF':
            tools.notifyLog('Turn 3D %s mode off' % (_currentMode), level=xbmc.LOGDEBUG)
            if _currentMode == 'SBS':
                __mode3D_off__ = __mode3DSBS_off__
            elif _currentMode == 'TAB':
                __mode3D_off__ = __mode3DTAB_off__
            else:
                return

            self.sendCommand(__mode3D_off__[self.lg_protocol], self.lg_seq_3D_off)
            self.isPlaying3D = False

    def poll(self):
        _host = None if __addon__.getSetting('lg_host') == '' else __addon__.getSetting('lg_host')
        try:
            tools.notifyLog('Service running (%s, %s)' % (self.lg_protocol, self.lg_pairing_key))

            while not self.Mon.abortRequested:
                _cycle = POLL_INTERVAL if self.sessionEstablished else WAIT_FOR_NEXT_CONNECT
                if self.Mon.waitForAbort(_cycle): break
                if self.Mon.settingsChanged: self.readSettings()
                if not self.sessionEstablished: self.getSettings(init=False)

        except interface.Interface.LGinNetworkNotFoundException:
            tools.notifyLog('LG Devices not found in network.', level=xbmc.LOGERROR)
            tools.dialogOSD( __LS__(30050))
        except interface.Interface.LGProtocolWebOSException:
            tools.notifyLog('Device use WebOS on port 3000. Not supported.', level=xbmc.LOGERROR)
            tools.dialogOSD(__LS__(30051))
        except interface.Interface.LGProtocollNotAcceptedException:
            tools.notifyLog('Protocol not supported.', level=xbmc.LOGERROR)
            tools.dialogOSD(__LS__(30052))
        except interface.Interface.NoConnectionToHostException:
            tools.notifyLog('No connection to host.', level=xbmc.LOGERROR)
            tools.dialogOSD(__LS__(30053) % (_host))
        except Exception, e:
            pass

RemoteService = Service()
RemoteService.poll()
del RemoteService
tools.notifyLog('Service finished')